User authentication design pattern overhaul (#258)
pxblue/doc-it: src/docs/patterns/user-auth.mdx

import ChangePassword from '../../app/assets/design-patterns/user-authentication/change-password.png';
import ChangePasswordWelcome from '../../app/assets/design-patterns/user-authentication/change-password-confirm.png';

<!-- Behaviors -->

import EmailCommunicationInvitation from '../../app/assets/design-patterns/user-authentication/email-communication-invitation.png';
import EmailCommunicationVerification from '../../app/assets/design-patterns/user-authentication/email-communication-verification.png';
import ErrorStates from '../../app/assets/design-patterns/user-authentication/error-states.png';

<!-- KEYWORDS: design pattern login registration register password forget forgot account authentication auth join -->

# User Authentication

<ImageGrid images={[LoginBanner]} regularWidth />

Most applications will require some form of user authentication to ensure that only authorized users have access.

<!-- @TODO switch login links over to the "Component Recommendations style" instead of an "Interactive Code example" once we have one designed. -->

<React.Fragment>
    <DemoCard repository={'login'} angular react ionic reactNative />
    <Divider />

> </React.Fragment>

Author: https://github.com/pxblue/react-workflows/issues/9 Need to solve this issue first

> </React.Fragment>

Reviewer: Can we switch it to the auth-workflow and comment it out for now?

> </React.Fragment>

Author: Sure. What about react native? It seems to not set up to run on snack

> </React.Fragment>

Reviewer: In addition to the upstream issues we're having with Snack/Expo, the demo for RN is built using the RNC CLI, not Expo, so it is not importable into Snack. We'll have to leave it out for now (maybe a future update we can tweak this component to be able to link to a github repo for some things that don't have a live demo but still ave demo code).

# Variations

<ImageGrid

>     images={[ViewPortLoginImage, ViewPortOverlayImage]}

Author: I try to use this paragraph to explain how they would look differently on mobile & desktop, and then I can present only the mobile screens from here on.

>     images={[ViewPortLoginImage, ViewPortOverlayImage]}

Author: Moved to the behavior section under the headline "Responsive Design"

>     images={[ViewPortLoginImage, ViewPortOverlayImage]}

Author: The purpose of those screens was to show (1) that we have several different auth-related workflows and (2) how they work on different viewport.

>     images={[ViewPortLoginImage, ViewPortOverlayImage]}

Reviewer: I get that. I was just a little confused because it's the first image under variations, but it's only showing 2 of the (6?).

If you leave it in, you should update the caption — it refers to all 4 as login screens.

        `1&2: An example of a login workflow, on mobile and on desktop.`,
        `3&4: An example of a password changing workflow, on mobile and on desktop`,
        `Note that #2 represents a typical desktop auth layout when the user is NOT logged in, while #4's layout is typical for the workflow when the user has already logged in.`,
    ]}
/>

Your user authentication workflow should work on different screen sizes. On mobile, it should take up an entire screen. On a desktop, it is usually "boxed" in a card. The content should follow the rules specified by our [forms and validation pattern](/patterns/forms), and each page should be connected using our [steppers pattern](/patterns/steppers).

<MaterialDesignDescription 
    title={'Forms and Validation'}
    url={'/patterns/forms'}
    description={`Follow PX Blue's guideline for form controls.`}
     avatar={<PxblueSmallAltIcon style={{ width: 48, height: 48 }} />}
/>
<MaterialDesignDescription 
    title={'Steppers'}
    url={'/patterns/steppers'}
    description={`Follow PX Blue's guideline for navigating between pages.`}
     avatar={<PxblueSmallAltIcon style={{ width: 48, height: 48 }} />}
/>

> Currently, PX Blue supports the following authentication workflows:

Author: I am hesitant to use the word "components" -- they mean something else in our usual context.

> Currently, PX Blue supports the following authentication workflows:

Reviewer: Functions? Modules? Elements? Sub-workflows? User paths? 

"Underneath the Authentication umbrella, there are four primary workflows:"

> Currently, PX Blue supports the following authentication workflows:

Author: Changed to "the user authentication contains four primary workflows"

-   Registration
-   Join organization
-   Login
-   Reset password

We will now discuss them one by one.

## Registration

Registration is the process of user account creation. Depends on your project setup, the user might need to request to join various organizations, which we will discuss later. Registration includes the following variations:

**Self-Registration**

Self-registration involves registering an account without an email invitation. The workflow involves the following screens:

-   Enter email address
-   Accept EULA
-   Verify email
-   Create password
-   Add account details
-   Success & welcome

**Registration via Invitation**

Sometimes the organization admin might email the user an invitation. The workflow involves the following screens:

-   Click through an email invitation
-   Accept EULA
-   Create password
-   Add account details

> -   Success & welcome

Author: No love for email invitation? It is quite critical and recurs in nearly every auth-related design prototypes.

> -   Success & welcome

Reviewer: It's not a screen, it's something that happens behind the scenes

> -   Success & welcome

Author: should I change the wording to "The invitation-based registration workflow includes the following **steps**" instead?

> -   Success & welcome

Author: Aha, saw your comments down below.

### Screens

> Most steps involved in self-registration and registration via invitation are very similar. Use the following screens for the registration workflow where appropriate.

Reviewer: ```suggestion
The majority of the registration screens are the same between invitation-based and self-registration. The primary difference is in how the registration workflow is initiated. Use the following screens for the registration workflow where appropriate.
```

> <ImageGrid

Reviewer: Should we show desktop version as well since they are a little different in terms of how the header looks?

Author: 🤔  I personally think the images at the very beginning of the grids section is good enough.

Reviewer: I would suggest mixing in a couple desktop ones to be more explicit and not force adopters to do too much extrapolation. The one at the beginning of the page (change password) is a single screen, so we don't currently have any examples of what a multi-screen workflow looks like on Desktop.

Reviewer: We don't need a desktop version for every image, but it might be nice to have a few sprinkled in or alternate, etc.

    images={[EnterEmailAdress, AcceptEULA, VerifyEmail, CreatePassword, AccountDetails, WelcomeOnboard]}
    caption={[
        'Screens shared by the registration workflows.',
        '1: Enter email address',
        '2: Accept EULA',
        '3: Verify email',
        '4: Create password',
        '5: Add account details',

>         '6: Success & welcome',

Reviewer: ```suggestion
        '6: Success / Welcome',
```

    ]}
/>

> ## Joining an Organization

Reviewer: Does this happen as part of registration, or is it presented later in the application after logging in?

Author: It is usually tightly connected with the registration, although people do usually have the option to join another organization on their own.

> In addition to account creation, some applications may require the user to join groups and organizations. Joining an organization workflow includes the following variations:

Reviewer: ```suggestion
Some applications may require users to join specific groups or organizations after creating an account. Generally, there are three ways for a user to join an organization:
```

**Request Access**

> This occurs when a user has an account but need to access an invitation-only organization. The workflow involves the following screens:

Reviewer: ```suggestion
Requesting access occurs when a user already has an account but needs to access an invitation-only organization. This workflow includes the following screens:
```

> -   Enter organization code
> -   Success & welcome (awaiting approval)
> -   Receive an email invitation
> -   Success & welcome

Reviewer: ```suggestion
-   Enter organization code
-   Request success
-   Joined success / welcome
```

Author: Should we rename all these Success / Welcome screens to: 

Success: request sent
Success: joined organization
Success: Account created

etc?

With a prefix, it is easier to tell which screens fall in that "success/welcome" category.

Author: And same as before, I think email invitation is an integrated part of this experience.

Reviewer: Yeah, I think the prefix would work.

I agree sending the email is a critical part of the flow, but it's not a screen. If you want to keep it in there, you could reword the introductory sentence to be something more like "The X workflow involves the following _steps_:"

**Enter Passcode**

> This occurs when anyone with an organization passcode can be added to the organization. The workflow involves the following screens:

Reviewer: ```suggestion
Joining by code occurs when anyone with an organization passcode can join an organization. This workflow includes the following screens:
```

Author: Changed to: 

> when anyone with an organization **identifier and** passcode

-   Enter organization code
-   Enter organization passcode

> -   Success & welcome

Reviewer: ```suggestion
-   Success / Welcome
```

**Join with Invitation**

> An organization admin can add an existing user from their end too. In case the user doesn't have an account, they will first do the registration -> registration via invitation, then proceed to this step. The workflow involves the following screens:

Reviewer: ```suggestion
Joining via invitation occurs when a system admin extends an invitation for a user to join a group. If the user does not yet have an account, they will first proceed through the invitation-based registration workflow followed by the screens for joining the organization. This workflow includes the following screens:
```

Author: Sometimes I feel like writing in kindergarten English bc I know Joe will always be backing me off

> -   Receive an email invitation
> -   Success & welcome

Reviewer: ```suggestion
-   Success / Welcome
```

### Screens

> Most steps involved in the joining-an-organization workflow are very similar. Use the following screens for the joining-an-organization workflow where appropriate.

Reviewer: ```suggestion
The majority of the joining-an-organization screens are the same regardless of the underlying mechanism. Use the following screens for the joining-an-organization workflow where appropriate.
```

<ImageGrid

>     images={[EnterOrganizationCode, EnterOrganizationPasscode, EnterOrganizationConfirm, EnterOrganizationWelcome]}

Reviewer: Same question as above...do we need to show desktop versions too?

    caption={[

>         'Screens shared by the joining-an-organization workflows.',
>         '1: Enter organization code',
>         '2: Enter organization passcode',
>         '3: Success & welcome (awaiting approval)',
>         '4: Success & welcome',

Reviewer: ```suggestion
        'Screens in the joining-an-organization workflow.',
        '1: Enter organization code',
        '2: Enter organization passcode',
        '3: Request success',
        '4: Joined success / welcome',
```

    ]}
/>

## Login

> Login includes the following variations:

Reviewer: ```suggestion
Logging in can be accomplished in one of two ways: basic login or login via a third-party.
```

**Basic Login**

> A simple login presents fields for the users to enter their credentials and a button to submit them for verification.

Reviewer: This is a section that I think could be fleshed out a bit. What are the extras? What does clicking them do? When do you show them?

Author: You mean the "extra functions" mentioned in the caption?

Reviewer: Yeah, like the "Forgot Password" (arguably not an extra, but rather required), "Contact Support", "Sign Up".

I'm just thinking like an extra 2 sentences below the image that quickly says something to the effect of "You should have a Forgot password button that will initiate the Forgot password Workflow (see below)"

> <ImageGrid
>     images={[Login]}
>     caption={[
>         'A standard login screen.',
>         '1: Project identifications, such as project name or project logo.',
>         '2: Optionaly, you may add in a simple geometrical background image with a white gradient background.',
>         '3: "Remember me" and "password visibility" are two features that facilitate the user, but expose some security concerns.',
>         '4: Login screens often include extra options, such as "forgot password", "sign up", and "contact your organization admin".',
>     ]}
> />

Reviewer: ```suggestion
<ImageGrid
    images={[Login]}
    caption={[
        'A standard login screen.',
        '1: Product identification, such as product name or logo.',
        '2: Optionally, you may add in a simple geometrical background image with a white gradient background.',
        '3: "Remember me" and "password visibility" are two features that facilitate the user, but expose some security concerns.',
        '4: Login screens often include extra functions, such as "forgot password", "sign up", and "contact your organization admin".',
    ]}
/>
```

> **Login through Third Party**

Reviewer: ```suggestion
**Login via Third Party**
```

> Some applications, especially the consumer-faced ones, are authorized via third-party credentials. Different platforms may have different requirements for this. Here are some commonly used platforms and their instructions:
> 
> -   [Login with Amazon](https://developer.amazon.com/docs/login-with-amazon/brand.html)
> -   [Google Sign-in](https://developers.google.com/identity/sign-in/web/sign-in)
> -   [Sign in with Apple](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/introduction/)
> -   [Facebook Login](http://developers.facebook.com/docs/facebook-login/)
> -   [Sign in with Microsoft](https://developer.microsoft.com/en-us/identity/add-sign-in-with-microsoft)

Reviewer: ```suggestion
Some applications, especially consumer-facing ones, can be authorized via third-party accounts / credentials. Different platforms may have different requirements for this. Here are some commonly used platforms and their instructions:

-   [Login with Amazon](https://developer.amazon.com/docs/login-with-amazon/brand.html)
-   [Google Sign-in](https://developers.google.com/identity/sign-in/web/sign-in)
-   [Sign in with Apple](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/introduction/)
-   [Facebook Login](http://developers.facebook.com/docs/facebook-login/)
-   [Sign in with Microsoft](https://developer.microsoft.com/en-us/identity/add-sign-in-with-microsoft)
```

## Reset Password

**Forget password**

> The login screen for an application should also present users with a way to recover or reset their password if they have lost or forgotten it. This is typically presented as a hyperlink below the credential fields. Clicking the link will present users with a field to enter their e-mail address. They will receive an email with a hyperlink to reset their password. The workflow involves the following screens:

Reviewer: ```suggestion
The login screen for an application should also present users with a way to recover or reset their password if they have lost or forgotten it. This is typically presented as a hyperlink below the username / password fields. Clicking the link will take the user to a Forgot Password screen where they can enter their e-mail address used to register. They will receive an email with a hyperlink to reset their password. This workflow includes the following screens:
```

> -   Enter email
> -   Success (email has been sent)
> -   Receive an email for resetting the password
> -   Create password
> -   Success & welcome

Reviewer: ```suggestion
-   Enter email address
-   Reset instructions sent
-   Create a new password
-   Success
```

<ImageGrid
    images={[ForgotPasswordEnterEmail, ForgotPasswordConfirmSent, ForgotPasswordNewPassword, ForgotPasswordWelcome]}

>     caption={['1: Enter email', '2: Success (Email has been sent)', '3: Create Password', '4: Success & Welcome']}

Reviewer: ```suggestion
    caption={['1: Enter email address', '2: Reset instructions sent', '3: Create a new password', '4: Success']}
```

/>

> **Change password**
> 
> When the users are logged in, they have the option to reset their password by going through the changing password workflow. This workflow can usually be triggered from user preference settings. The workflow involves the following screens:
> 
> -   Change password
> -   Success & welcome

Reviewer: ```suggestion
**Change password**

When the users are logged in, they have the option to update or change their password by going through the change password workflow, typically triggered from within the user settings. This workflow includes the following screens:

-   Change password
-   Success
```

Author: >the option to update or change their password

what would be the difference between "update" and change" in this context?

Reviewer: Effectively no difference. We can just say change.

I think in my head I may have been making a slight distinction between being required to update your password (based on an IT policy) and voluntarily changing your password. But I think change is fine to cover both scenarios.

<React.Fragment>
    <ImageGrid
        images={[ChangePassword, ChangePasswordWelcome]}

>         caption={['1: Change password', '2: Success & welcome']}

Reviewer: ```suggestion
        caption={['1: Change password', '2: Success']}
```

    />
    <Divider />
</React.Fragment>

# Behavior

## Email Communications

<!-- This might deserve its own pattern. We already have something in the backlog to design for an email template. -->

> Email communications are frequently used throughout the user authentication workflow since they typically involve some security changes. You should never give the users option to unsubscribe from account-related email communications.

Reviewer: Is this more about sending emails when account information changes, or about the emails associated with the Authentication / Registration workflows (i.e., sending the codes, invites, etc.)?

Author: I was trying to emphasize that for the sake of the account security, so in that regard, it is closer to the first point you made. Yet it is so frequently mentioned throughout the steps above, so I think it is worthwhile to bring it up in this pattern

Reviewer: I would probably break it into two sections (one for each type):

1. Emails are often sent as part of the authentication workflows ... blah blah blah. They include invitations to register, confirmation codes, etc.

2. Emails are also sent to confirm security-related changes to a user's account (such as changing a password, updating a profile, etc.)...blah blah blah. You should never allow users to unsubscribe from security-related email notifications.

<ImageGrid
    images={[EmailCommunicationInvitation, EmailCommunicationVerification]}
    caption={['Two examples of email communications', '1: Invitation to join', '2: Verifying email address']}
/>

## Success & Welcome Screen

<!-- Welcome screen might deserve to be its own pattern (or at least merge into something like "onboarding"). -->
<!-- TODO: add more images here, if studio blue can release more. -->

> The purpose of a welcome screen is to assure the user that something has been indeed achieved, and thus can be more delightful. You are encouraged to use animations and [graphics](/style/illustrations).

Reviewer: ```suggestion
The purpose of a welcome screen is to assure the user that something has been achieved. These screens can be an opportunity to perform more product branding — you are encouraged to use animations and [graphics](/style/illustrations) on these screens.
```

## Terminated Workflows & Expired Sessions

During the registration, if a user exits the application before they ever create a password for their account, the whole session is treated as expired, and the users should start over again.

> If an invitation link or verification code has expired, the users should be able to request for another invitation link or verification code.

Reviewer: Could use a little more fleshing out. Doesn't mention expired sessions / session timeout at all. Images?

Author: Let me invent some then, probably with a dialog box

## Security Concerns

### Error Messages

When a user encounters an error, whether it is a password mismatch or the server is temporarily down, they should be notified with an appropriate error message. However, be very careful when presenting error messages to the users. You should only show what is necessary to recover from the problem, such as "incorrect email or password" and "server unavailable".

> 

Reviewer 2: the show what's necessary (error message) shouldn't indicate incorrect email or incorrect password. Can we use some of the verbiage from original login doc for this section.

If authentication fails, the user remains on the login page and the screen should be updated to show an appropriate error message (e.g., Invalid Credentials).

If an error occurs when attempting to validate the credentials (e.g., Server Unavailable), the user should be notified with an appropriate error message.

- NOTE: be very careful when presenting error messages to users. For security, you should not present more information than is necessary (e.g., do NOT provide messages like "User does not exist," "Incorrect Password," etc.).

Author: @jeffvg @joebochill may you guys clarify what exactly shall I change?

Reviewer 2: for this section - You should only show what is necessary to recover from the problem, such as "incorrect email or password" and "server unavailable".

maybe something like
- You should only show what is necessary to recover from the problem, such as "invalid credentials and "server unavailable".

NOTE: be very careful when presenting error messages to users. For security, you should not present more information than is necessary (e.g., do NOT provide messages like "User does not exist," "Incorrect Password," etc.).

Author: @jeffvg "incorrect email or password" is a phrase I grabbed from WAS. Would you suggest me to change that to "invalid credentials", in the text and in the illustrations?

Reviewer 2: I would say yes. Normally we wouldn't want to call out what part of the credentials are incorrect. If password is incorrect, both ID and password fields would error "invalid credentials". Keeping hackers away....lol

<ImageGrid
    images={[ErrorStates]}
    caption={
        'For security, you should not present more information than is necessary. For invalid credentials, you should highlight both fields and say "incorrect email or password."'
    }
    regularWidth
/>

> ### Credential Leaks
> 
> Unlike a phone application, a shared HMI application has more threats in credential leaks. These projects should grant their user neither the ability to toggle their password's visibility nor should the application remember users' credentials (i.e., "Remember Me").

Reviewer: ```suggestion
### Exposing Credentials

Unlike a phone application, a shared HMI application is more vulnerable to outside threats via showing credentials. These projects should **NOT** implement password visibility toggles or "remember me" functionality.
```

### Two-Factor Authentication

The examples on this page used email as the login credentials. However, if your project seeks for additional securities, consider using two-factor authentication.

<Divider />

> # Developer

Reviewer: Are we limiting these components to just the workflow packages, or should we also include some of the basics if they want to roll their own:
Input, Card, etc.?

Author: I am inclined to picture them as "we recommend using these pxb packages"

Those inputs & cards should already be mentioned in form pattern & overlay pattern, so I don't think it is worth repeating them here.

Reviewer: I think that's fine. I would just match the same verbiage we use in the other developer sections "You should use the following packages to implement this pattern:"

And then maybe link out to the Forms and Overlays patterns again. Something like:

"Additionally, you should familiarize yourself with the components used for the Forms and Overlays patterns".

**Angular:**

_Planned_

**React:**

-   [@pxblue/react-auth-workflow](https://www.npmjs.com/package/@pxblue/react-auth-workflow)

**React Native:**

> PX Blue has an npm package ([@pxblue/react-native-auth-workflow](https://www.npmjs.com/package/@pxblue/react-native-auth-workflow)) for React Native. It includes pre-built implementations of the screens for Login, Forgot Password, Contact Information, Self-Registration, Registration By Invitation, and Change Password.

Reviewer: Match the format of the other ones, we don't need all of the description here.

import ForgotPasswordWelcome from '../../app/assets/design-patterns/user-authentication/forgot-password-welcome.png';
import ChangePassword from '../../app/assets/design-patterns/user-authentication/change-password.png';
import ChangePasswordWelcome from '../../app/assets/design-patterns/user-authentication/change-password-confirm.png';
import ResetPasswordDesktop from '../../app/assets/design-patterns/user-authentication/reset-password-desktop.png';
import ChangePasswordDesktop from '../../app/assets/design-patterns/user-authentication/change-password-desktop.png';

<!-- Behaviors -->

import EmailCommunicationInvitation from '../../app/assets/design-patterns/user-authentication/email-communication-invitation.png';
import EmailCommunicationVerification from '../../app/assets/design-patterns/user-authentication/email-communication-verification.png';
import ExpiredCode from '../../app/assets/design-patterns/user-authentication/expired.png';
import ErrorStates from '../../app/assets/design-patterns/user-authentication/error-states.png';

<!-- KEYWORDS: design pattern login registration register password forget forgot account authentication auth join -->

> # User Authentication & Registration

Reviewer: Should we just call this thing "Login & Registration"? Will that fit on one line in the drawer?

Reviewer: Maybe not. "User Authentication" in the sidebar doesn't seem like it covers everything that is in here (registration). It's probably fine though.

<ImageGrid images={[LoginBanner]} regularWidth />

> Most applications will require some form of user authentication to ensure that only authorized users have access as well as a mechanism for users to be added to the application.

Reviewer: This is a little long-winded for an intro sentence. Maybe:
```suggestion
Most applications require some form of user authentication (to keep unauthorized users out) and a way to add authorized users to the application.
```

<!--

@TODO switch login links over to the "Component Recommendations style" instead of an "Interactive Code example" once we have one designed.

<React.Fragment>
    <DemoCard repository={'login'} angular react={'https://codesandbox.io/s/github/pxblue/react-workflows/tree/master/login-workflow/example'} ionic reactNative />
    <Divider />
</React.Fragment>

-->

<Divider />

# Variations

> The workflows should follow the rules specified by our [forms and validation](/patterns/forms) pattern, and multi-step workflows should be connected using our [steppers](/patterns/steppers) pattern.

Reviewer: ```suggestion
All of these workflows should follow the rules specified by our [forms and validation](/patterns/forms) pattern, and multi-step workflows should be connected using our [steppers](/patterns/steppers) pattern.
```

<MaterialDesignDescription 
    title={'Forms and Validation'}
    url={'/patterns/forms'}
    description={`Follow PX Blue's guideline for form controls.`}
     avatar={<PxblueSmallAltIcon style={{ width: 48, height: 48 }} />}
/>
<MaterialDesignDescription 
    title={'Steppers'}
    url={'/patterns/steppers'}
    description={`Follow PX Blue's guideline for navigating between pages.`}
     avatar={<PxblueSmallAltIcon style={{ width: 48, height: 48 }} />}
/>

Your authentication and registration workflows should work equally well on different screen sizes. On mobile, these should be presented as full-screen layouts, whereas on desktop, they should be presented in a card centered on the screen as an [overlay](/patterns/overlay).

> The user authentication contains four primary workflows:

Reviewer: ```suggestion
User authentication contains four primary workflows:
```

-   Receive an email invitation
-   Success: organization joined

### Screens

The majority of the joining-an-organization screens are the same regardless of the underlying mechanism. Use the following screens for the joining-an-organization workflow where appropriate.

<ImageGrid
    images={[EnterOrganizationCode, EnterOrganizationPasscode, EnterOrganizationConfirm, EnterOrganizationWelcome]}
    caption={[
        'Screens in the joining-an-organization workflow.',
        '1: Enter organization identifier',
        '2: Enter organization passcode',
        '3: Success: request sent',
        '4: Success: organization joined',
    ]}
/>
<ImageGrid images={[JoinAnOrganizationDesktop]} caption={'A join-an-organization screen example on desktop.'} regularWidth />

## Login

<!--
An NFC login might either be documented here, or wait until we start adding in device registery / discovery pattern
-->

Logging in can usually be accomplished in one of two ways: basic login or login via a third-party.

**Basic Login**

A simple login presents fields for the users to enter their credentials and a button to submit them for verification.

> The login screen should always offer some workaround if a user has some trouble getting authorized, either because they don't have an account, or that they forget their credentials. In your login screen, there should be buttons that can initiate those extra functions, such as Registration and Forgot Password (see below).

Reviewer: ```suggestion
The login screen should always present information on how to create an account and what to do if a user has forgotten their password — these buttons should trigger the Registration and Forgot Password (see below) workflows .

> If you are using invitation-based registration, you should still provide information on how to request an invite.
```

Author: Although we currently don't have designs for it yet — should we take into account the fact that the user might forget their email as well?

Reviewer: Sure, we can say "Credentials" instead of password to cover both

<ImageGrid
    images={[Login]}
    caption={[
        'A standard login screen.',
        '1: Product identifications, such as product name or logo.',
        '2: Optionally, you may add in a simple geometrical background image with a white gradient background.',

>         '3: "Remember me" and "password visibility" are two features that facilitate the user, but expose some security concerns.',

Reviewer: ```suggestion
        '3: "Remember me" and "password visibility" are two features that improve user experience, but introduce potential security concerns.',
```

        '4: Login screens often include extra functions, such as "forgot password", "sign up", and "contact your organization admin".',
        '5: You may add certification badges and logos from partnerships here.',
    ]}
    gridImageProps={{ sm: 8 }}
/>
<ImageGrid images={[LoginDesktop]} caption={'A login screen example on desktop.'} regularWidth />

**Login via Third Party**

Some applications, especially consumer-facing ones, can be authorized via third-party accounts / credentials. Different platforms may have different requirements for this. Here are some commonly used platforms and their instructions:

-   [Login with Amazon](https://developer.amazon.com/docs/login-with-amazon/brand.html)
-   [Google Sign-in](https://developers.google.com/identity/sign-in/web/sign-in)
-   [Sign in with Apple](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/introduction/)
-   [Facebook Login](http://developers.facebook.com/docs/facebook-login/)
-   [Sign in with Microsoft](https://developer.microsoft.com/en-us/identity/add-sign-in-with-microsoft)

> ## Forgot / Reset Password

Reviewer: ```suggestion
## Forgot / Change Password
```

Author: Does "reset password" include both forget password and change password?

Reviewer: I think Forgot covers Forgot/Reset Password and
Change Password covers change password.

Forgot/Reset is slightly different from Change because Change happens when you are already authenticated whereas Reset has an extra step of sending the email.

> **Forget password**

Reviewer: ```suggestion
**Forgot Password**
```

> The login screen for an application should also present users with a way to recover or reset their password if they have lost or forgotten it. This is typically presented as a hyperlink below the credential fields. Clicking the link will present users with a field to enter their e-mail address. They will receive an email with a hyperlink to reset their password. This workflow involves the following steps:

Reviewer: ```suggestion
The login screen for an application should also present users with a way to reset their password if they have forgotten it. This is typically displayed as a hyperlink below the credential fields. Clicking the link will present users with a field to enter their e-mail address. They will then receive an email with a hyperlink to reset their password. This workflow involves the following steps:
```

-   Enter email address
-   Success: reset instructions sent
-   Receive an email for resetting the password
-   Create a new password

> -   Success: password reset

Reviewer: ```suggestion
-   Success: password updated
```

<ImageGrid
    images={[ForgotPasswordEnterEmail, ForgotPasswordConfirmSent, ForgotPasswordNewPassword, ForgotPasswordWelcome]}
    caption={['1: Enter email address', '2: Success: reset instructions sent', '3: Create a new assword', '4: Success: password reset']}
/>
<ImageGrid images={[ResetPasswordDesktop]} caption={'A reset-password screen example on desktop.'} regularWidth />

> **Change password**

Reviewer: ```suggestion
**Change Password**
```

> When the users are logged in, they have the option to change their password by going through the change password workflow, typically triggered within the user settings. This workflow includes the following steps:

Reviewer: ```suggestion
When the users are logged in, they have the option to change their password by going through the change password workflow, typically triggered from within the user settings. This workflow includes the following steps:
```

-   Change password
-   Success: password changed

> 

Reviewer: ```suggestion
> After a successful password change, users should be logged out and forced to log in with the new credentials.
```

Author: good catch

<React.Fragment>
    <ImageGrid
        images={[ChangePassword, ChangePasswordWelcome]}
        caption={['1: Change password', '2: Success: password changed']}
    />
    <ImageGrid images={[ChangePasswordDesktop]} caption={'A change-password screen example on desktop.'} regularWidth />
    <Divider />
</React.Fragment>

# Behavior

## Email Communications

<!-- This might deserve its own pattern. We already have something in the backlog to design for an email template. -->

> Email communications are frequently used throughout the user authentication workflow, as they are the primary way a user will get notified without an account access. Users receive emails for invitations, verification, confirmation codes, etc.

Reviewer: ```suggestion
Email communications are frequently used throughout the user authentication workflow to verify the user's identity. Users receive emails for invitations, address verification, confirmation codes, etc.
```

> Emails are also sent to confirm security-related changes to a user's account (such as changing a password, updating a profile, etc.) You should never allow users to unsubscribe from security-related email notifications.

Reviewer: ```suggestion
Emails are also sent to confirm security-related changes to a user's account, such as changing a password, updating a profile, etc. You should never allow users to unsubscribe from security-related email notifications.
```

<ImageGrid
    images={[EmailCommunicationInvitation, EmailCommunicationVerification]}
    caption={['Two examples of email communications', '1: Invitation to join', '2: Verifying email address']}
/>

> ## Success Screen

Reviewer: Does this belong in Behaviors? Or should it be somewhere further up?

Author: I understand what makes you say so. I put it there just because it feels cleaner.

Reviewer: Is it short enough to maybe put it in a 
> callout ?

<!-- Welcome screen might deserve to be its own pattern (or at least merge into something like "onboarding"). -->
<!-- TODO: add more images here, if studio blue can release more. -->

The purpose of a welcome screen is to assure the user that something has been achieved. These screens can be an opportunity to perform more product branding — you are encouraged to use animations and [graphics](/style/illustrations) on these screens.

## Terminated Workflows & Expired Sessions

> During the registration, if a user exits the application before they ever create a password for their account, the whole session is treated as expired, and the users should start over again.

Reviewer: ```suggestion
During the registration, if a user leaves the application before creating a password for their account, the whole session is treated as expired, and upon returning, they must start the registration process over again.
```

If an invitation link or verification code has expired, the users should be able to request for another invitation link or verification code.

<ImageGrid
    images={[ExpiredCode]}
    caption={'Users should have a way to re-initiate an expired email verification.'}
    regularWidth
/>

## Security Concerns

### Error Messages

> When a user encounters an error, whether it is a password mismatch or the server is temporarily down, they should be notified with an appropriate error message. However, be very careful when presenting error messages to the users. You should only show what is necessary to recover from the problem, such as "incorrect email or password" and "server unavailable".

Reviewer: ```suggestion
When a user encounters an error — such as a password mismatch or the server is temporarily down — they should be notified with an appropriate error message. However, be very careful when presenting error messages to the users. You should only show what is necessary to recover from the problem, such as "incorrect email or password" or "server unavailable" — do not expose information that could compromise the security of your application.
```

> > Do **NOT** provide messages like "User does not exist," "Incorrect Password," etc.

Reviewer: "Did you forgot the % sign at the end of your password, Peter Horowitz?"

<ImageGrid
    images={[ErrorStates]}
    caption={
        'For security, you should not present more information than is necessary. For invalid credentials, you should highlight both fields and say "Invalid Credentials."'
    }
    regularWidth
/>

### Exposing Credentials

Unlike a phone application, a shared HMI application is more vulnerable to outside threats via showing credentials. These projects should **NOT** implement password visibility toggles or "remember me" functionality.

### Two-Factor Authentication

> The examples on this page used email as the login credentials. However, if your project seeks for additional securities, consider using two-factor authentication.

Reviewer: ```suggestion
The examples on this page used email as the login credentials. However, if your project requires additional security, consider using two-factor authentication. Two-factor authentication involves sending an additional confirmation code (typically via email or SMS) to verify a user's identity after they enter their password.
```
